### PR TITLE
BONNIE-1455: hybrid measure support to display SDE

### PR DESCRIPTION
**Pull request by c-monkey**
Do not display any SDEs unless measure is specified as hybrid at upload. Otherwise, display all SDEs. This PR also tests the bonnie_bundler change that added the hybrid field: https://github.com/projecttacoma/bonnie_bundler/pull/148

Pull requests into Bonnie require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

------
- [ ] **Need to update fixture `CMS529v0` on bonnie.healthit.gov** which can't be added until PR deployed. Then re-export and update fixture in this PR. Use the new version of the Lookback measure in BONNIE-1408: (https://jira.mitre.org/secure/attachment/27515/CCDELookback_v5_4_Artifacts%20%2813%29.zip)
------
- [x] bonnie_bundler reference updated
------

**Submitter:**
- [X] This pull request describes why these changes were made.
- [X] This PR is into the correct branch.
- [X] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1455
- [x] JIRA ticket links to this PR
- [X] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [X] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [X] Tests are included and test edge cases
- [X] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [X] Test fixtures updated and documented as necessary ( see [internal wiki](https://gitlab.mitre.org/bonnie/internal-documentation/wikis/testing#test-fixtures) )
- [X] Code coverage has not gone down and all code touched or added is covered. 
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here: 
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here: 

Branch | Back End Coverage | Front End Coverage
-- | -- | --
bonnie-v2.2 | 1738 / 1950 LOC (89.13%) | Statements   : 58.94% (7134/12104)<br>Branches     : 42.79% ( 1935/4522 )<br>Functions    : 49.19% ( 1178/2395 )<br>Lines        : 56.81% ( 6408/11279 )
bonnie_1454_1455_hybrid_SDE | 1739 / 1951 LOC (89.13%) | Statements   : 58.97% (7138/12104)<br>Branches     : 43.06% ( 1947/4522 )<br>Functions    : 49.21% ( 1178/2394 )<br>Lines        : 56.85% ( 6412/11279 )

- [X] Automated regression test(s) pass
    * No measures with differences
    * Several measures had covered clauses now uncovered (SDE Ethnicity, Race, Sex, and Payer):
        * rachel@oregonurology.com - CMS771v1
        * epecqmmpr@gmail.com - CMS460v0
        * hmdmnewmeasures@gmail.com - CMS812v2
        * hmdmtjc2@gmail.com - CMS334v1 and CMS334v1

If JIRA tests were used to supplement or replace automated tests:
- [X] JIRA test links: N/A
- [X] Justification for using JIRA tests: N/A
- [X] JIRA tests have been added to sprint N/A

**Reviewer 1:**

Name: @mayerm94 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [X] JIRA tests have been run and pass N/A
- [X] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

**Reviewer 2:**

Name: @jbradl11 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [X] JIRA tests have been run and pass N/A
- [X] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

### CODE REVIEW

File: app/assets/javascripts/templates/import/import_measure.hbs [hunk 1]
```diff
@@ -176,6 +177,13 @@
                   <a class="btn btn-primary"></a>
                 </div>
               </div>
+              <div class="col-sm-offset-{{titleSize}} col-sm-{{subtract dataSize 1}}">
+                  <div class="checkbox">
+                      <label>
+                          <input id="hybridCheckbox" type="checkbox" name="hybrid_measure" value="true" {{#if isHybrid}}checked{{/if}}/>Hybrid measure
```
Comment (mayerm94) [inline app/assets/javascripts/templates/import/import_measure.hbs]: This checkbox looks out of place, like it was haphazardly appended to the view. One option would be to match the pattern in place IE: Option title on the left and then the selectable box, to denote non-hybrid and hybrid with the non-hybrid measure option selected by default. Another option would be do put the checkbox under type, either way I think the option should fall under one of the categories defined by the left most titles

Comment (lukoktonos) [reply app/assets/javascripts/templates/import/import_measure.hbs]: We discussed this a bit and tried out a few different UI options and think this might be the way to go:
image

Using this wording makes it more natural to be placed underneath the Calculations section and might be more descriptive to measure developers.

What do you think, @edeyoung?

Comment (c-monkey) [reply app/assets/javascripts/templates/import/import_measure.hbs]: That seems great. Are we going to do anything else besides including SDEs to differentiate from a "non-hybrid" measure?

File: app/assets/javascripts/templates/measure.hbs [hunk 2]
```diff
@@ -33,6 +33,9 @@
         {{/unless}}
         <p></p>
       {{/if}}
+      {{#if hybrid}}
+          <p>Hybrid Measure</p>
```
Comment (jbradl11) [inline app/assets/javascripts/templates/measure.hbs]: I am thinking this label should match the "upload option" that is discussed above. It could be confusing to users if we display "Hybrid Measure" if the upload dialog doesn't mention it.

File: lib/tasks/bonnie.rake [hunk 3]
```diff
@@ -185,7 +185,11 @@ namespace :bonnie do
       end
     end
 
-    desc 'Downloads a measure package for a particular measure'
+    desc %{Downloads a measure package for a particular measure.
+
+    Provide either HQMF_SET_ID or CMS_ID
+
+    $ rake bonnie:db:download_measure_package EMAIL=bonnie-fixtures@mitre.org HQMF_SET_ID=8C942F13-6EAA-4A8E-BB4F-3B3BF7311C1F CMS_ID=761}
```
Comment (jbradl11) [inline lib/tasks/bonnie.rake:192]: Is this an unrelated code cleanup change?

File: app/assets/javascripts/templates/import/import_measure.hbs (hunk 1 above)
Comment (edeyoung) [reply app/assets/javascripts/templates/import/import_measure.hbs]: @losborne i think what you are showing makes sense

File: app/assets/javascripts/templates/measure.hbs (hunk 2 above)
Comment (c-monkey) [reply app/assets/javascripts/templates/measure.hbs]: Yes, now that we're all on the same page, the text does need to agree

File: lib/tasks/bonnie.rake (hunk 3 above)
Comment (c-monkey) [reply lib/tasks/bonnie.rake:192]: good point, just expanding the comment on the rake task

File: app/assets/javascripts/templates/import/import_measure.hbs [hunk 4]
```diff
@@ -176,6 +177,13 @@
                   <a class="btn btn-primary"></a>
                 </div>
               </div>
+              <div class="col-sm-offset-{{titleSize}} col-sm-{{subtract dataSize 1}}">
+                  <div class="checkbox">
+                      <label>
+                          <input id="hybridCheckbox" type="checkbox" name="sde_calc" value="true" {{#if calcSDEs}}checked{{/if}}/>Include Supplemental Data Element Calculations
```
Comment (mayerm94) [inline app/assets/javascripts/templates/import/import_measure.hbs]: Suggest changing ID of checkbox in the same way all other 'hybrid' language was changed

File: app/assets/javascripts/helpers/cql_results_helpers.js.coffee [hunk 5]
```diff
@@ -60,7 +60,7 @@ class CQLResultsHelpers
       for statementName of statements
         statementRelevance[lib][statementName] = "NA"
 
-    if CQLMeasureHelpers.isHybridMeasure(measure)
+    if measure.get('calculate_sdes')
       for statement in populationSet.get('supplemental_data_elements')
```
Comment (jbradl11) [inline app/assets/javascripts/helpers/cql_results_helpers.js.coffee]: I understand this is not part of this PR. However, this line breaks Bonnie when uploading older measures that are uploaded with the checkbox `Include Supplemental Data Element Calculations` checked.

File: app/controllers/measures_controller.rb [hunk 6]
```diff
@@ -60,7 +60,8 @@ def create
 
     measure_details = {
      'type'=>params[:measure_type],
-     'episode_of_care'=>params[:calculation_type] == 'episode'
+     'episode_of_care'=>params[:calculation_type] == 'episode',
+     'calculate_sdes'=>params[:sde_calc]
```
Comment (jbradl11) [inline app/controllers/measures_controller.rb]: Seems like everywhere else you are using variables named "calcSDEs" or "calculate_sdes" its not a priority but I would suggest changing this to `:calc_sde`

Comment (c-monkey) [reply app/controllers/measures_controller.rb]: fine by me!

File: app/assets/javascripts/helpers/cql_results_helpers.js.coffee (hunk 5 above)
Comment (c-monkey) [reply app/assets/javascripts/helpers/cql_results_helpers.js.coffee]: @jbradl11 Will you please let me know which measure was failing? I'd like to fix this bug.

File: app/assets/javascripts/templates/import/import_measure.hbs (hunk 4 above)
Comment (c-monkey) [reply app/assets/javascripts/templates/import/import_measure.hbs]: Good catch. Thanks! Fixed.

File: app/assets/javascripts/helpers/cql_results_helpers.js.coffee (hunk 5 above)
Comment (c-monkey) [reply app/assets/javascripts/helpers/cql_results_helpers.js.coffee]: So, the null check I added fixes this issue and JB and I decided the existing warning message is sufficient: i.e. "Outdated Measure! This measure was written using an outdated version of QDM. Please re-package and re-export the measure from the MAT."

The other choice we decided against is preventing upload, which seems too stringent.

**Comment by c-monkey** [comment]
Please review despite the brakeman error. I don't think that's appropriate to fix in this PR. We can fix it first for the master branch and then integrate it in, perhaps?

**Review by jbradl11 — APPROVED** [review]
PR looks good. Gemfile will need to be updated and travis is currently breaking, however that is a known issue.

**Review by mayerm94 — APPROVED** [review]

**Comment by jbradl11** [comment]
Although the commit history seems a little odd, the Code diff looks good.
